File: iclusion-importer/src/main/java/com/hartwig/hmftools/iclusion/api/IclusionApiWrapper.java
package com.hartwig.hmftools.iclusion.api;

import java.util.List;
import java.util.concurrent.TimeUnit;

import com.squareup.moshi.Moshi;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import okhttp3.ConnectionPool;
import okhttp3.Dispatcher;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;

public final class IclusionApiWrapper {

    private static final Logger LOGGER = LogManager.getLogger(IclusionApiWrapper.class);

    private IclusionApiWrapper() {
    }

    public static void connectWithIclusionApi(@NotNull String iClusionEndpoint, @NotNull String iClusionClientId,
            @NotNull String iClusionClientSecret, @NotNull String iClusionUsername, @NotNull String iClusionPassword) {
        LOGGER.info("Connecting with iClusion API on {}", iClusionEndpoint);

        Dispatcher dispatcher = new Dispatcher();
        dispatcher.setMaxRequests(100);
        dispatcher.setMaxRequestsPerHost(100);

        OkHttpClient client = new OkHttpClient.Builder().connectionPool(new ConnectionPool(20, 5, TimeUnit.SECONDS))
                .readTimeout(20, TimeUnit.SECONDS)
                .connectTimeout(20, TimeUnit.SECONDS)
                .writeTimeout(20, TimeUnit.SECONDS)
                .dispatcher(dispatcher)
                .build();

        MoshiConverterFactory moshiConverterFactory =
                MoshiConverterFactory.create(new Moshi.Builder().add(new IclusionResponseAdapter()).build());

        Retrofit retrofit = new Retrofit.Builder().baseUrl(iClusionEndpoint)
                .addConverterFactory(moshiConverterFactory)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
                .client(client)
                .build();

        IclusionApi api = retrofit.create(IclusionApi.class);

        MultipartBody requestBody = new MultipartBody.Builder().setType(MultipartBody.FORM)
                .addFormDataPart("grant_type", "password")
                .addFormDataPart("client_id", iClusionClientId)
                .addFormDataPart("client_secret", iClusionClientSecret)
                .addFormDataPart("username", iClusionUsername)
                .addFormDataPart("password", iClusionPassword)
                .build();

        String tokenBearer = "Bearer " + api.requestAccessToken(requestBody).blockingFirst().accessToken;
        System.out.println("TokenBearer = " + tokenBearer);

        List<IclusionStudy> studies = api.studies(tokenBearer).blockingFirst();
        System.out.println("Studies = " + studies.size());

        List<IclusionIndication> indications = api.indications(tokenBearer).blockingFirst();
        System.out.println("Indications = " + indications.size());

        List<IclusionGene> genes = api.genes(tokenBearer).blockingFirst();
        System.out.println("Genes = " + genes.size());

        List<IclusionVariant> variants = api.variants(tokenBearer).blockingFirst();
        System.out.println("Variants = " + variants.size());

        client.dispatcher().executorService().shutdown();
    }
}

File: api-clients/src/main/kotlin/com/hartwig/hmftools/apiclients/iclusion/api/IclusionApiWrapper.kt
package com.hartwig.hmftools.apiclients.iclusion.api

import com.hartwig.hmftools.apiclients.iclusion.data.*
import com.hartwig.hmftools.apiclients.iclusion.http.httpClient
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import io.reactivex.Observable
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.LogManager
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory

class IclusionApiWrapper(endpoint: String, private val clientId: String, private val clientSecret: String,
                         private val user: String, private val password: String) {
    companion object {
        private fun createApi(httpClient: OkHttpClient, endpoint: String): IclusionApi {
            val retrofit = Retrofit.Builder().baseUrl(endpoint)
                    .addConverterFactory(moshiConverter())
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
                    .client(httpClient)
                    .build()
            return retrofit.create(IclusionApi::class.java)
        }

        private fun moshiConverter(): MoshiConverterFactory {
            return MoshiConverterFactory.create(Moshi.Builder().add(KotlinJsonAdapterFactory()).add(IclusionResponseAdapter).build())
        }
    }

    private val logger = LogManager.getLogger("IclusionApiWrapper")

    private val httpClient = httpClient()
    private val api = createApi(httpClient, endpoint)
    private val tokenBearer = "Bearer ${getAccessToken().blockingFirst().access_token}"

    private fun getAccessToken(): Observable<Token> {
        val requestBody = MultipartBody.Builder().setType(MultipartBody.FORM)
                .addFormDataPart("grant_type", "password")
                .addFormDataPart("client_id", clientId)
                .addFormDataPart("client_secret", clientSecret)
                .addFormDataPart("username", user)
                .addFormDataPart("password", password)
                .build()
        return api.getAccessToken(requestBody)
    }

    fun studyDetails(): List<IclusionStudyDetails> {
        val studies = studies().blockingIterable().toList()
        logger.info("Queried ${studies.size} studies via iclusion API.")

        logger.info(" Studies with CCMO identifier: ${studies.filterNot { it.ccmo.isEmpty() }.size}")
        logger.info(" Studies with mutations configured: ${studies.filterNot { it.mutations.isEmpty() }.size}")
        logger.info(" Studies with EUDRA identifier: ${studies.filterNot { it.eudra.isEmpty() }.size}")
        logger.info(" Studies with NCT identifier: ${studies.filterNot { it.nct.isEmpty() }.size}")
        logger.info(" Studies with IPN ?: ${studies.filterNot { it.ipn.isNullOrEmpty() }.size}")

        studies.filterNot { it.ipn.isNullOrEmpty() }.forEach { logger.info(it) }

        val indications = indications().blockingIterable().toList().associateBy { it.id }
        logger.info("Queried ${indications.size} indications via iclusion API.")

        val genes = genes().blockingIterable().toList().associateBy { it.id }
        logger.info("Queried ${genes.size} genes via iclusion API.")

        val variants = variants().blockingIterable().toList().associateBy { it.id }
        logger.info("Queried ${variants.size} variants via iclusion API.")

        return studies.filterNot { it.mutations.isEmpty() }.map {
            val indicationsForStudy = it.indication_ids.mapNotNull { indications[it] }
            val mutationsDetails = it.mutations.mapNotNull {
                val geneName = genes[it.gene_id]?.gene_name
                val variantName = variants[it.variant_id]?.variant_name
                geneName ?: return@mapNotNull null
                variantName ?: return@mapNotNull null
                return@mapNotNull IclusionMutationDetails(geneName, variantName)
            }
            IclusionStudyDetails(it, indicationsForStudy, mutationsDetails)
        }.filterNot { it.mutations.isEmpty() }
    }

    fun close() {
        httpClient.dispatcher().executorService().shutdown()
    }

    private fun indications(): Observable<IclusionIndication> {
        return api.indications(tokenBearer).flatMapIterable { it }
    }

    private fun genes(): Observable<IclusionGene> {
        return api.genes(tokenBearer).flatMapIterable { it }
    }

    private fun variants(): Observable<IclusionVariant> {
        return api.variants(tokenBearer).flatMapIterable { it }
    }

    private fun studies(): Observable<IclusionStudy> {
        return api.studies(tokenBearer).flatMapIterable { it }
    }
}
